File: packages/contracts/test/go/conversation_work_test.go
package contracts_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"

	contracts "convenewire.dev/contracts/generated/go"
	runtimecontracts "convenewire.dev/contracts/generated/go/runtime"
)

func TestConversationWireInterop(t *testing.T) {
	raw, err := os.ReadFile(filepath.Join(packageRoot(t), "fixtures", "conversation-work-cases.json"))
	if err != nil {
		t.Fatal(err)
	}
	var suite struct {
		Cases []struct {
			Name     string
			Valid    bool
			Instance json.RawMessage
		}
	}
	if json.Unmarshal(raw, &suite) != nil {
		t.Fatal("invalid fixture")
	}
	for _, item := range suite.Cases {
		t.Run(item.Name, func(t *testing.T) {
			err := runtimecontracts.ValidateBridgeMessage(item.Instance)
			if (err == nil) != item.Valid {
				t.Fatalf("valid=%t error=%v", item.Valid, err)
			}
			if !item.Valid {
				return
			}
			var envelope struct {
				Type string `json:"type"`
			}
			_ = json.Unmarshal(item.Instance, &envelope)
			var value any = &contracts.RunRequestedMessage{}
			if envelope.Type == "run.reply" {
				value = &contracts.RunReplyMessage{}
			}
			if err := json.Unmarshal(item.Instance, value); err != nil {
				t.Fatal(err)
			}
			encoded, err := json.Marshal(value)
			if err != nil || !reflect.DeepEqual(decodeJSON(t, item.Instance), decodeJSON(t, encoded)) {
				t.Fatalf("wire round trip changed the proposal: %s %v", encoded, err)
			}
		})
	}
}

File: packages/contracts/test/conversation-work.test.mjs
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import test from "node:test";
import { validateBridgeMessage } from "../generated/runtime/bridge-validator.mjs";

test("conversation proposals retain a bounded optional wire contract", async () => {
  const suite = JSON.parse(await readFile(new URL("../fixtures/conversation-work-cases.json", import.meta.url), "utf8"));
  for (const item of suite.cases) assert.equal(validateBridgeMessage(item.instance), item.valid, item.name);
  const legacy = structuredClone(suite.cases[0].instance);
  delete legacy.payload.conversationWork;
  assert.equal(validateBridgeMessage(legacy), true);
});
